### PR TITLE
feat(compiler): make `keySpan` available for `BoundAttributes`

### PR DESCRIPTION
**Pull request by atscott**
Though we currently have the knowledge of where the `key` for an
attribute binding appears during parsing, we do not propagate this
information to the output AST. This means that once we produce the
template AST, we have no way of mapping a template position to the key
span alone. The best we can currently do is map back to the
`sourceSpan`. This presents problems downstream, specifically for the
language service, where we cannot provide correct information about a
position in a template because the AST is not granular enough.

### CODE REVIEW

File: packages/compiler/test/render3/r3_ast_spans_spec.ts [hunk 1]
```diff
@@ -237,9 +239,9 @@ describe('R3 AST source spans', () => {
       // </ng-template>
       expectFromHtml('<div *ngFor="let item of items"></div>').toEqual([
         ['Template', '0:38', '0:32', '32:38'],
-        ['TextAttribute', '6:11', '<empty>'],  // ngFor
-        ['BoundAttribute', '5:31', '25:30'],   // *ngFor="let item of items" -> items
-        ['Variable', '13:22', '<empty>'],      // let item
+        ['TextAttribute', '6:11', '<empty>'],          // ngFor
+        ['BoundAttribute', '5:31', '22:24', '25:30'],  // *ngFor="let item of items" -> of -> items
```
Comment (petebacondarwin) [inline packages/compiler/test/render3/r3_ast_spans_spec.ts:243]: Is `of` really the correct key value span for this binding?
I would have expected it to be `*ngFor`.

File: packages/compiler/src/expression_parser/ast.ts [hunk 2]
```diff
@@ -841,7 +841,8 @@ export class ParsedProperty {
 
   constructor(
       public name: string, public expression: ASTWithSource, public type: ParsedPropertyType,
-      public sourceSpan: ParseSourceSpan, public valueSpan?: ParseSourceSpan) {
+      public sourceSpan: ParseSourceSpan, readonly keySpan?: ParseSourceSpan,
+      public valueSpan?: ParseSourceSpan) {
```
Comment (petebacondarwin) [inline packages/compiler/src/expression_parser/ast.ts:845]: I assume you are 100% sure you have checked all places that call this and for the similar classes below? Since the type checker will not fail if you have passed a `valueSpan` in the slot for the new `keySpan`.

File: packages/compiler/test/render3/r3_ast_spans_spec.ts (hunk 1 above)
Comment (atscott) [reply packages/compiler/test/render3/r3_ast_spans_spec.ts:243]: `of` is technically the correct `keySpan` for the microsyntax. The de-sugared form of this is `ngFor let-item [ngForOf]="items"` -- the `of items` parts is the section in the microsyntax that maps to the `ngForOf`. This actually illustrates one of the problems that was difficult to solve when trying to go map backwards from the template AST to the template for the Language Service. If given the location for `of`, there wasn't anywhere in the AST that had that span specifically.

File: packages/compiler/src/expression_parser/ast.ts [hunk 3]
```diff
@@ -841,7 +841,8 @@ export class ParsedProperty {
 
   constructor(
       public name: string, public expression: ASTWithSource, public type: ParsedPropertyType,
-      public sourceSpan: ParseSourceSpan, public valueSpan?: ParseSourceSpan) {
+      public sourceSpan: ParseSourceSpan, readonly keySpan: ParseSourceSpan|undefined,
```
Comment (kyliau) [inline packages/compiler/src/expression_parser/ast.ts:844]: +1 for making it readonly!

Please document when `keySpan` is undefined and when it's not.

File: packages/compiler/src/template_parser/binding_parser.ts [hunk 4]
```diff
@@ -206,7 +206,7 @@ export class BindingParser {
   parseLiteralAttr(
       name: string, value: string|null, sourceSpan: ParseSourceSpan, absoluteOffset: number,
       valueSpan: ParseSourceSpan|undefined, targetMatchableAttrs: string[][],
-      targetProps: ParsedProperty[]) {
+      targetProps: ParsedProperty[], keySpan?: ParseSourceSpan) {
```
Comment (kyliau) [inline packages/compiler/src/template_parser/binding_parser.ts:209]: Please add a doc here to explain `keySpan` is not passed in VE AST implementation.

File: packages/compiler/src/template_parser/binding_parser.ts [hunk 5]
```diff
@@ -216,18 +216,19 @@ export class BindingParser {
...
-          ParsedPropertyType.LITERAL_ATTR, sourceSpan, valueSpan));
+          ParsedPropertyType.LITERAL_ATTR, sourceSpan, keySpan, valueSpan));
     }
   }
 
   parsePropertyBinding(
       name: string, expression: string, isHost: boolean, sourceSpan: ParseSourceSpan,
       absoluteOffset: number, valueSpan: ParseSourceSpan|undefined,
-      targetMatchableAttrs: string[][], targetProps: ParsedProperty[]) {
+      targetMatchableAttrs: string[][], targetProps: ParsedProperty[], keySpan?: ParseSourceSpan) {
```
Comment (kyliau) [inline packages/compiler/src/template_parser/binding_parser.ts:231]: same as comment above, please add doc

File: packages/compiler/test/render3/r3_ast_spans_spec.ts [hunk 6]
```diff
@@ -144,28 +146,28 @@ describe('R3 AST source spans', () => {
     it('is correct for bound properties', () => {
       expectFromHtml('<div [someProp]="v"></div>').toEqual([
         ['Element', '0:26', '0:20', '20:26'],
-        ['BoundAttribute', '5:19', '17:18'],
+        ['BoundAttribute', '5:19', '6:14', '17:18'],
```
Comment (kyliau) [inline packages/compiler/test/render3/r3_ast_spans_spec.ts:149]: Please add a trailing comment to denote what sourceSpan, keySpan, and valueSpan are, like what we've done for some of the newer tests.

File: packages/compiler/src/render3/r3_ast.ts [hunk 7]
```diff
@@ -43,12 +43,17 @@ export class BoundAttribute implements Node {
...
       public name: string, public type: BindingType, public securityContext: SecurityContext,
       public value: AST, public unit: string|null, public sourceSpan: ParseSourceSpan,
-      public valueSpan?: ParseSourceSpan, public i18n?: I18nMeta) {}
+      readonly keySpan: ParseSourceSpan, public valueSpan: ParseSourceSpan|undefined,
+      public i18n: I18nMeta|undefined) {}
 
-  static fromBoundElementProperty(prop: BoundElementProperty, i18n?: I18nMeta) {
+  static fromBoundElementProperty(prop: BoundElementProperty, i18n?: I18nMeta): BoundAttribute
+      |null {
+    if (prop.keySpan === undefined) {
```
Comment (kyliau) [inline packages/compiler/src/render3/r3_ast.ts:51]: If keySpan is undefined here, we should just throw an error, because it indicates a bug in our implementation, not user code.

On a separate note, doesn't this mean we'll return null for `BoundElementProperty` for host binding?
I think the solution we discussed is to use sourceSpan as the keySpan for this edge case.
If possible, please look into how we could add a test for this scenario.

File: packages/compiler/src/template_parser/binding_parser.ts [hunk 8]
```diff
@@ -243,40 +244,41 @@ export class BindingParser {
...
-          sourceSpan, valueSpan, targetMatchableAttrs, targetProps);
+          sourceSpan, keySpan, valueSpan, targetMatchableAttrs, targetProps);
     }
   }
 
   parsePropertyInterpolation(
       name: string, value: string, sourceSpan: ParseSourceSpan,
       valueSpan: ParseSourceSpan|undefined, targetMatchableAttrs: string[][],
-      targetProps: ParsedProperty[]): boolean {
+      targetProps: ParsedProperty[], keySpan?: ParseSourceSpan): boolean {
```
Comment (kyliau) [inline packages/compiler/src/template_parser/binding_parser.ts:259]: ditto

File: packages/compiler/src/render3/r3_template_transform.ts [hunk 9]
```diff
@@ -332,15 +338,23 @@ class HtmlAstToIvyAst implements html.Visitor {
     const absoluteOffset =
         attribute.valueSpan ? attribute.valueSpan.start.offset : srcSpan.start.offset;
 
+    function createKeySpan(srcSpan: ParseSourceSpan, prefix: string, identifier: string) {
+      const normalizationAdjustment = attribute.name.length - name.length;
```
Comment (kyliau) [inline packages/compiler/src/render3/r3_template_transform.ts:342]: Please add comment to explain that this is to account for `data-`.

File: packages/compiler/src/render3/r3_template_transform.ts [hunk 10]
```diff
@@ -332,15 +338,23 @@ class HtmlAstToIvyAst implements html.Visitor {
...
+      const keySpanStart = srcSpan.start.moveBy(prefix.length + normalizationAdjustment);
+      const keySpanEnd = keySpanStart.moveBy(identifier.length + normalizationAdjustment);
```
Comment (kyliau) [inline packages/compiler/src/render3/r3_template_transform.ts:344]: we don't need normalization adjustment for keySpanEnd since it's relative to keySpanStart, no?
If that's the case, I wonder why our tests didn't catch this...

File: packages/compiler/src/render3/r3_template_transform.ts [hunk 11]
```diff
@@ -315,7 +315,13 @@ class HtmlAstToIvyAst implements html.Visitor {
         // input of a directive and directive matching happens at runtime.
         const bep = this.bindingParser.createBoundElementProperty(
             elementName, prop, /* skipValidation */ true, /* mapPropertyName */ false);
-        bound.push(t.BoundAttribute.fromBoundElementProperty(bep, i18n));
+        const attribute = t.BoundAttribute.fromBoundElementProperty(bep, i18n);
+        if (attribute !== null) {
+          bound.push(attribute);
+        } else {
+          this.reportError(
```
Comment (kyliau) [inline packages/compiler/src/render3/r3_template_transform.ts:322]: We shouldn't need this, since reportError() is for errors in user code.

File: packages/compiler/src/render3/r3_template_transform.ts [hunk 12]
```diff
@@ -332,15 +338,23 @@ class HtmlAstToIvyAst implements html.Visitor {
...
+      return new ParseSourceSpan(keySpanStart, keySpanEnd, identifier);
+    }
+
     const bindParts = name.match(BIND_NAME_REGEXP);
     let hasBinding = false;
 
     if (bindParts) {
       hasBinding = true;
       if (bindParts[KW_BIND_IDX] != null) {
+        const keySpan = createKeySpan(srcSpan, bindParts[KW_BIND_IDX], bindParts[IDENT_KW_IDX]);
```
Comment (kyliau) [inline packages/compiler/src/render3/r3_template_transform.ts:354]: Nit: In the else cases below, `bindParts[IDENT_KW_IDX]` is extracted into a variable named `identifier`. We should do the same here.

**Review by kyliau — CHANGES_REQUESTED** [review]

File: packages/compiler/test/render3/r3_ast_spans_spec.ts (hunk 6 above)
Comment (atscott) [reply packages/compiler/test/render3/r3_ast_spans_spec.ts:149]: Resolving as superseded by https://github.com/angular/angular/pull/38902

File: packages/compiler/src/render3/r3_template_transform.ts (hunk 10 above)
Comment (atscott) [reply packages/compiler/src/render3/r3_template_transform.ts]: Yikes, good catch! I added tests for all the data- binding types

File: packages/compiler/src/render3/r3_ast.ts (hunk 7 above)
Comment (atscott) [reply packages/compiler/src/render3/r3_ast.ts:50]: Yes, this would return null (actually error now after the change) for host bindings. However, `BoundElementProperty`s of host bindings aren't converted to a `BoundAttribute`s but rather converted to a host binding function here: https://github.com/angular/angular/blob/88d7bb838607622c48488a828286c54fe091d8f0/packages/compiler/src/render3/view/compiler.ts#L546. I've added a lengthy comment to `binding_parser` about using `sourceSpan` as the `keySpan` for host bindings.

**Review by kyliau — APPROVED** [review]
LGTM, this looks great!

**Review by alxhub — APPROVED** [review]
Wow, this was definitely more work than I expected.

Consider making this a `refactor()` commit as it's not a user-facing feature and should not appear in the CHANGELOG.
